lib/modules: add mkUnset, mkOverrideUnset

rhendric opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

Using only `mkForce`/`mkOverride`, a module or configuration has no way
of removing an attribute specified elsewhere; the best you could do is
to replace that attribute with a null, empty set, or similar. This may
result in undesired behavior, if a module implementation takes some
action for every attribute in a set, for example.

`mkOverrideUnset` is a function intended to be used like `mkOverride`: a
`config.some.attribute.path = lib.mkOverrideUnset p` means that, if `p`
is the winning priority among all definitions of `some.attribute.path`,
then `some.attribute.path` will not exist at all in the final
configuration values.

`mkUnset` is a shorthand for `mkOverrideUnset 50`, analogous to
`mkForce`.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

netixx commented:
Can't say anything about the implementation, but the API seems good to me :)

Very much like to have this feature as well!

danbst reviewed (approved):
@rhendric  kool!

A few thoughts:
- this works awkward with default values - if option has some _default_, mkUnset will remove it (but restoring default value isn't that hard, as we can query `options`)
- attrs inside lists are impossible to unset. (those are also impossible to `mkForce`, so no big deal)

But this definitely has it's uses.

Infinisil commented:
Do you have some concrete examples of where this would have been really useful? In general I think attrsets having an `.enable` option is nicer than this, but I understand that not all attrset options provide that, so this might be needed after all.

One thing I'd like to see for this is a test in `nixpkgs/lib/tests` though.

danbst commented:
@Infinisil this is mostly useful for `attrsOf XXX` option types, which are easy to add new ones, but impossible to remove existing ones. Luckily we have `systemd.services.*.enable`, but we don't have `containers.*.enable` or `services.nginx.virtualHosts.*.enable`,

rhendric commented:
@Infinisil, there's also things like `systemd.services.*.unitConfig.*`, which, once set, can't be individually removed without clobbering the entire `unitConfig` (you can set them individually to `null`, but that still leaves an empty `Foo=` line in the generated unit file). I suspect `.enable` wouldn't be a palatable way to handle individual `unitConfig` entries. I wonder if this pattern—using an attrset of configuration options in which an absent attribute induces a different outcome than an attribute that is present with some neutral value—appears elsewhere in nixpkgs... if not, maybe this is just an edge case.

> One thing I'd like to see for this is a test in `nixpkgs/lib/tests` though.

Ah, that's where they live! Yes, I'll gladly do that.

Infinisil commented:
Cool, sounds reasonable to add this then. I'm also going to ping @nbp and @rycee.

rycee commented on nixos/doc/manual/release-notes/rl-1909.xml:
```diff
@@ -297,6 +297,11 @@
    </listitem>
+   <listitem>
+    <para>
+     The <literal>mkOverrideUnset</literal> function and <literal>mkUnset</literal> constant have been added to <literal>lib</literal>.
```
In `option-def.xml` the `mkOverrideUnset` and `mkUnset` names were inside `varname` elements. Might be nice to continue that here?

rhendric replied:
Okay, done!

rycee commented on nixos/doc/manual/development/option-def.xml:
```diff
@@ -74,6 +74,12 @@ services.openssh.enable = mkOverride 10 false;
+  <para>
+   The function <varname>mkOverrideUnset</varname> can be used instead to
+   remove an attribute that was defined in another module. It also takes a
+   priority, but no replacement value. <varname>mkUnset</varname> is equal to
```
Generally I would avoid starting a sentence with a piece of code. Perhaps something like
```suggestion
   priority, but no replacement value. For convenience, <varname>mkUnset</varname> is equal to
```
?

rhendric replied:
Sure, done!

rycee reviewed (commented):
I think this is a very nice idea and as far as I can tell the implementation is fine. I would like to see a few tests under `lib/tests/modules`, though. Both to help verify that the implementation behaves as expected and to avoid future regressions.

Edit: I notice now that @Infinisil already asked for tests 🙂

rhendric committed:
lib/modules: add mkUnset, mkOverrideUnset

Using only `mkForce`/`mkOverride`, a module or configuration has no way
of removing an attribute specified elsewhere; the best you could do is
to replace that attribute with a null, empty set, or similar. This may
result in undesired behavior, if a module implementation takes some
action for every attribute in a set, for example.

`mkOverrideUnset` is a function intended to be used like `mkOverride`: a
`config.some.attribute.path = lib.mkOverrideUnset p` means that, if `p`
is the winning priority among all definitions of `some.attribute.path`,
then `some.attribute.path` will not exist at all in the final
configuration values.

`mkUnset` is a shorthand for `mkOverrideUnset 50`, analogous to
`mkForce`.

rycee reviewed (approved):
Looks good to me!

rhendric commented:
Ping, @nbp? (Or whoever is the appropriate person to pull the lever—apologies if I got it wrong.)

nbp commented:
I have not yet looked into the code of the patch, but I will try to look at it by the end of the weekend, feel free to remind me over irc if I forgot.

Multiple comments:
 - the proper wording is undefined, and `mkUndef`. `Unset` implies an imperative order of execution, which does not exist here. Instead, you define an empty value that take precedence or not over other values which are defined.
 - This has nothing to do with `mkOverride`, this provide a non-existing value which then can be combined with `mkOverride` and `mkIf`. `mkOVerrideUndef` is literally equivalent to `mkOverride … mkUndef`.
 - This used to exists and got removed, and we should investigate why before merging these changes.

nbp commented:
`mkNotDef` got removed by https://github.com/NixOS/nixpkgs/commit/0e333688cea468a28516bf6935648c03ed62a7bb#diff-490491facef0aa6a6dbe8ffc5bd97097 , probably because the properties managed to be implemented without this attribute.

There is not justification specific to `mkNotDef`, only arguments about the efficiency of the module system.

nbp reviewed (changes requested):
This code does not push down the `unset` property down to any attributes within, for the simple reason that there is no attribute within.

The problem is that overriden properties are evaluated at the location of option declaration, and as such one cannot use `mkUnset` above a set of attribute to unset, like `mkOverride` or `mkIf`. Therefore the following test case would not work with the proposed patch:

```
{ options.foo.bar = mkOption { type = attrsOf int; … }
  config = mkMerge {
    (mkUnset)
    ({ foo = mkUnset; })
    # ({ foo.bar = mkUnset; })
    # ({ foo.bar.alpha = mkUnset; })
    ({ foo.bar.alpha = 1; })
  };
}
```

dali99 commented:
Something like this is still imporant to me

RaitoBezarius commented:
Just ran into a case where I needed this again.

rhendric commented:
I think something like this is still probably a good idea, and I understand my own code well enough, but I don't understand the surrounding system well enough to properly address what review feedback I got, and I'm no longer maintaining the NixOS configuration I had that wanted this. So I'm going to close this PR and I explicitly grant permission to any other interested party to use any or all of this code as a starting point for a new one.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/from-a-nixos-module-is-it-possible-to-delete-an-entry-from-an-attrset-defined-by-a-different-module/73289/9
